jobfunnel/config/cli.py

```diff
@@ -0,0 +1,408 @@
+       args_dict['master_csv_file'] or args_dict['cache_folder']
+       or args_dict['block_list_file'] or args_dict['duplicates_list_file']
+    )
+
+    # Build a config that respects CLI, defaults and YAML
+    # NOTE: we a passed settings YAML first so we can inject CLI after if needed
+    if settings_yaml_file:
+
+        # Load YAML
+        config.update(
+            yaml.load(open(settings_yaml_file, 'r'), Loader=yaml.FullLoader)
+        )
+        # Set defaults for our YAML
+        config = SettingsValidator.normalized(config)
+
+        # Validate the config passed via YAML
+        if not SettingsValidator.validate(config):
+            raise ValueError(
+                f"Invalid Config settings yaml:\n{SettingsValidator.errors}"
+            )
+
+    if output_folder:
+
+        # We must build paths based on -o path exclusively
+        if user_passed_paths:
+
+            # This is an invalid case
+            raise ValueError(
+                "Cannot combine -o with -csv, -blf, -cache, -dlf arguments, "
+                "as -o defines these paths."
+            )
+
+        elif settings_yaml_file:
+
+            # This is another invalid case
+            raise ValueError(
+                "Cannot combine -s YAML and -o argument, all file paths must "
+                "be specified individually."
+            )
+
+        else:
+            # Set paths based on passed output_folder directly:
+            config['master_csv_file'] = os.path.join(
+                output_folder, 'master.csv'
+            )
+            config['cache_folder'] = os.path.join(
+                output_folder, '.cache'
+            )
+            config['block_list_file'] = os.path.join(
+                config['cache_folder'], 'block.json'
+            )
+            config['duplicates_list_file'] = os.path.join(
+                config['cache_folder'], 'duplicates.json'
+            )
+            if args_dict['log_file']:
+                config['log_file'] = args_dict['log_file']
+            else:
+                config['log_file'] = os.path.join(
+                    output_folder, 'log.log'
+                )
+
+    elif user_passed_paths:
+
+        # Handle CLI arguments for paths, possibly overwriting YAML
+        for path_arg in PATH_ATTRS:
+            if args_dict.get(path_arg):
+                config[path_arg] = args_dict[path_arg]
+    else:
+
+        # NOTE: we will do this so that we can run without any arguments passed
+        output_folder = DEFAULT_OUTPUT_DIRECTORY
+
+    # Handle all the sub-configs, and non-path, non-default CLI args
+    for key, arg_value in args_dict.items():
+        if arg_value:
+            if key == 'log_level' and arg_value != DEFAULT_LOG_LEVEL_NAME:
+                # We got a non-default log level, overwrite any YAML setting
+                config[key] = arg_value
+            else:
+                # Set sub-config value
+                for sub_key in ['search', 'delay', 'proxy']:
```

Comment:
We still have the issue of the YAML file being overwritten by the CLI defaults, including the locale. I agree with you that we need to make the YAML file and CLI mutually exclusive; if we get a YAML file, we don't even look at the CLI defaults. If we get a CLI argument, the user _must_ pass all of the required CLI arguments and we enforce not allowing the user to pass in a YAML file. 

This will not only make our lives easier as developers, but it might actually make JobFunnel easier to use for users. Because right now is kind of fuzzy as to whether CLI args take precedence over YAML settings, or vice versa. This, I think, will make JobFunnel simpler to use in general.

Reply:
Agreed - ok, i will implement them as mutually exclusive option groups.